p2p/nat/natupnp.go

```diff
@@ -90,36 +114,51 @@ func (n *upnp) internalAddress() (net.IP, error) {
 }
 
 func (n *upnp) DeleteMapping(protocol string, extport, intport int) error {
-	return n.client.DeletePortMapping("", uint16(extport), strings.ToUpper(protocol))
+	return n.withRateLimit(func() error {
+		return n.client.DeletePortMapping("", uint16(extport), strings.ToUpper(protocol))
+	})
 }
 
 func (n *upnp) String() string {
 	return "UPNP " + n.service
 }
 
+func (n *upnp) withRateLimit(fn func() error) error {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	lastreq := time.Since(n.lastReqTime)
+	if lastreq < rateLimit {
+		time.Sleep(rateLimit - lastreq)
+	}
+	err := fn()
```

Comment:
Why are you passing through an error result here? The input params and the rest of the results are passed through the caller's local context, it just feels illogical to pass an error result here when you are not doing anything else with it. It could be a simple `fn func()`.

Reply:
I see, sometimes the only result is an error. If you still prefer it this way then I am fine with it, it just gave me a wrong first impression that the error result might have a special meaning to the rate limiter function.